src/nfa/operations.cc

```diff
@@ -56,6 +56,126 @@ namespace {
+            for (size_t j = 0; j < no_states; j++) {
+                matrix[i][j] = static_cast<size_t *>(malloc(sizeof(size_t) * no_states));
+            }
+        }
+
+        reverted_nfa = revert(aut); // Reverted NFA
+        // ! End of preprocessing
+
+        // ! Initial refinement
+        for (size_t x = 0; x < alph_syms.size(); x++) {
+            for (size_t p = 0; p < no_states; p++) {
+                for (size_t q = 0; q < no_states; q++) {
+                    size_t q_size;
+                    size_t p_size;
+
+                    auto symbol_q = aut.delta[q].find(alph_syms[x]);
+                    if (symbol_q == aut.delta[q].end()) {
+                        matrix[x][p][q] = 0;
+                        q_size = 0;
+                    } else {
+                        q_size = (*symbol_q).num_of_targets();
+                        matrix[x][p][q] = q_size;
+                    }
+
+                    auto symbol_p = aut.delta[p].find(alph_syms[x]);
+                    if (symbol_q == aut.delta[p].end()) {
+                        p_size = 0;
+                    } else {
+                        p_size = (*symbol_p).num_of_targets();
+                    }
+                    if ((p_size != 0 && q_size == 0) || (aut.final.contains(p) && !aut.final.contains(q))) {
+                        if (!vec_contains(result_sim, std::pair(p,q))) {
+                            result_sim.push_back(std::pair(p,q)); // R append 
+                            worklist.push_back(std::pair(p,q)); // worklist append
+                        }
+                    }
+                }
+            }
+        }
+        // ! End of initional refinement
+
+        // ! Propagate until fixpoint
+        size_t worklist_size;
+        std::pair<State, State> working_pair;
+        while ((worklist_size = worklist.size()) != 0) {
+            working_pair = worklist[worklist_size - 1];
+            worklist.pop_back();
+
+            for (size_t x = 0; x < alph_syms.size(); x++) {
+                auto symbol_q_ = reverted_nfa.delta[working_pair.second].find(alph_syms[x]);
+                if (symbol_q_ == reverted_nfa.delta[working_pair.second].end()) {
+                    continue;
+                } 
+                for (State q: (*symbol_q_).targets.to_vector()) {
+                    matrix[x][working_pair.first][q]--;
+                    if (matrix[x][working_pair.first][q] == 0) {
+                        auto symbol_p_ = reverted_nfa.delta[working_pair.first].find(alph_syms[x]);
+                        if (symbol_p_ == reverted_nfa.delta[working_pair.first].end()) {
+                            continue;
+                        }
+                        for (State p: (*symbol_p_).targets.to_vector()) {
+                            if (!vec_contains(result_sim, std::pair(p,q))) {
+                                result_sim.push_back(std::pair(p,q)); // R append 
```

Comment:
Instead of `push_back(std::pair(...))`, one can use `emplace_back(...)` to construct the pair in-place at the location in the vector.